File: app/ios/NeuroPulse/Protocol/NPProtocolLibrary.swift
import Foundation
import Combine

// MARK: - Protocol Availability

enum ProtocolAvailability: Equatable {
    case available
    case unavailable(missingModalities: [NPModalityType])

    var isAvailable: Bool {
        if case .available = self { return true }
        return false
    }

    var unavailableReason: String? {
        switch self {
        case .available:
            return nil
        case .unavailable(let missing):
            if missing.isEmpty { return "No device connected." }
            let names = missing.map { $0.displayName }.joined(separator: ", ")
            return "Requires: \(names)"
        }
    }
}

// MARK: - NPProtocolLibrary

@MainActor
final class NPProtocolLibrary: ObservableObject {

    // MARK: Published state

    @Published var userProtocols: [NPProtocolEntry] = []
    @Published var availableModalities: Set<NPModalityType> = NPModalityType.t1Set
    @Published var has1064SmartModules: Bool = false
    @Published var connectedDeviceTier: DeviceTier = .none

    // MARK: Device tier

    enum DeviceTier: Equatable {
        case none
        case t1
        case t2
    }

    // MARK: Persistence URL

    private static var persistenceURL: URL {
        URL.documentsDirectory.appending(path: "np_user_protocols.json")
    }

    // MARK: Init

    init() {
        loadFromDisk()
    }

    // MARK: Computed all-protocols list

    var allProtocols: [NPProtocolEntry] {
        NPPredefinedProtocols.all + userProtocols
    }

    // MARK: Availability

    func availability(for entry: NPProtocolEntry) -> ProtocolAvailability {
        guard connectedDeviceTier != .none else {
            return .unavailable(missingModalities: [])
        }

        // For composite protocols, resolve through the library
        if case .composite(let comp) = entry {
            var missingNames: Set<NPModalityType> = []
            for layer in comp.layers {
                if let proto = findSingle(named: layer.protocolName) {
                    let sub = availability(for: .single(proto))
                    if case .unavailable(let m) = sub { missingNames.formUnion(m) }
                }
            }
            return missingNames.isEmpty ? .available : .unavailable(missingModalities: Array(missingNames))
        }

        // Single protocol
        let required = entry.requiredModalityTypes

        // Check 1064nm smart module requirement
        var missing: [NPModalityType] = []
        if case .single(let proto) = entry, proto.requires1064SmartModule(), !has1064SmartModules {
            // Not strictly a missing modality but we flag the PBM modality
            // We treat it as unavailable pbmTranscranial (smart module variant)
            missing.append(.pbmTranscranial)
        }

        for modality in required {
            if !isModalityAvailable(modality) {
                if !missing.contains(modality) {
                    missing.append(modality)
                }
            }
        }

        return missing.isEmpty ? .available : .unavailable(missingModalities: missing)
    }

    private func isModalityAvailable(_ modality: NPModalityType) -> Bool {
        switch modality.tier {
        case .t1:
            return connectedDeviceTier == .t1 || connectedDeviceTier == .t2
        case .t2:
            return connectedDeviceTier == .t2
        case .accessory:
            // Accessories always show as provisionally available
            return true
        }
    }

    private func findSingle(named name: String) -> NPProtocolDefinition? {
        for entry in allProtocols {
            if case .single(let proto) = entry, proto.name == name { return proto }
        }
        return nil
    }

    // MARK: Mutations

    func save(_ entry: NPProtocolEntry) {
        let id = entry.id
        if let idx = userProtocols.firstIndex(where: { $0.id == id }) {
            userProtocols[idx] = entry
        } else {
            userProtocols.append(entry)
        }
        saveToDisk()
    }

    func delete(_ id: UUID) {
        userProtocols.removeAll { $0.id == id }
        saveToDisk()
    }

    // MARK: Script export / import

    func exportScript(_ entry: NPProtocolEntry) -> String {
        NPPSSerializer().serialize(entry)
    }

    func importScript(_ text: String) throws -> NPProtocolEntry {
        var lexer = NPPSLexer(text)
        let tokens = try lexer.tokenize()
        var parser = NPPSParser(tokens)
        let entries = try parser.parse()
        guard let first = entries.first else {
            throw NPPSError(message: "No protocol found in script", line: 0)
        }
        return first
    }

    // MARK: Device availability update

    func updateAvailability(tier: DeviceTier, smartModules: Bool) {
        connectedDeviceTier = tier
        has1064SmartModules = smartModules
        switch tier {
        case .none:
            availableModalities = []
        case .t1:
            availableModalities = NPModalityType.t1Set
        case .t2:
            availableModalities = NPModalityType.t2Set
        }
        if smartModules {
            availableModalities.insert(.pbmTranscranial) // already in, but belt+braces
        }
    }

    // MARK: Persistence

    private func loadFromDisk() {
        let url = Self.persistenceURL
        guard FileManager.default.fileExists(atPath: url.path) else { return }
        do {
            let data = try Data(contentsOf: url)
            let decoder = JSONDecoder()
            decoder.dateDecodingStrategy = .iso8601
            userProtocols = try decoder.decode([NPProtocolEntry].self, from: data)
        } catch {
            // If loading fails (e.g. schema migration), start with empty user protocols
            userProtocols = []
        }
    }

    private func saveToDisk() {
        let url = Self.persistenceURL
        do {
            let encoder = JSONEncoder()
            encoder.dateEncodingStrategy = .iso8601
            encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
            let data = try encoder.encode(userProtocols)
            try data.write(to: url, options: .atomic)
        } catch {
            // Silently fail; SHDR could log this in a real implementation
        }
    }
}

File: app/android/core/src/main/kotlin/life/neurone/core/protocol/NPProtocolLibrary.kt
package life.neurone.core.protocol

import life.neurone.core.common.KeyValueStore
import java.util.UUID

// Port of iOS NPProtocolLibrary (app/ios/NeurOne/Protocol/NPProtocolLibrary.swift).
// Pure-JVM: bundled + user protocol store, device-tier/EEG-consent availability, and
// validation integration. On iOS this is an @MainActor ObservableObject; on Android the
// UI layer observes it (a ViewModel wraps it). Persistence is via KeyValueStore using the
// NPPS text format (round-tripped through the existing serializer/parser) instead of iOS's
// JSON file — the on-device format is an implementation detail, behavior is identical.

// MARK: - Modality tier helpers (port of NPModalityType.tier / t1Set / t2Set) -----------

enum class NPModalityTier { T1, T2, ACCESSORY }

val NPModalityType.tier: NPModalityTier
    get() = when (this) {
        NPModalityType.PBM_TRANSCRANIAL, NPModalityType.PBM_INTRANASAL,
        NPModalityType.EEG_NEUROFEEDBACK, NPModalityType.BES_TACS, NPModalityType.TDCS,
        NPModalityType.VNS_HRV, NPModalityType.AUDIO_ENTRAINMENT, NPModalityType.VISUAL_STIMULATION ->
            NPModalityTier.T1
        NPModalityType.QEEG_21CH, NPModalityType.TMS, NPModalityType.PBM_DEEP_1170NM,
        NPModalityType.CLINICAL_TACS, NPModalityType.HD_TDCS, NPModalityType.CERVICAL_VNS ->
            NPModalityTier.T2
        NPModalityType.VIBROTACTILE_40HZ -> NPModalityTier.ACCESSORY
    }

val NP_T1_MODALITY_SET: Set<NPModalityType> = setOf(
    NPModalityType.PBM_TRANSCRANIAL, NPModalityType.PBM_INTRANASAL, NPModalityType.EEG_NEUROFEEDBACK,
    NPModalityType.BES_TACS, NPModalityType.TDCS, NPModalityType.VNS_HRV,
    NPModalityType.AUDIO_ENTRAINMENT, NPModalityType.VISUAL_STIMULATION,
)

val NP_T2_MODALITY_SET: Set<NPModalityType> = NP_T1_MODALITY_SET + setOf(
    NPModalityType.QEEG_21CH, NPModalityType.TMS, NPModalityType.PBM_DEEP_1170NM,
    NPModalityType.CLINICAL_TACS, NPModalityType.HD_TDCS, NPModalityType.CERVICAL_VNS,
)

// MARK: - NPProtocolEntry / NPProtocolDefinition computed helpers ------------------------

val NPProtocolEntry.id: UUID
    get() = when (this) {
        is NPProtocolEntry.Single -> protocol.id
        is NPProtocolEntry.Composite -> composite.id
        is NPProtocolEntry.Limits -> limits.id
    }

val NPProtocolEntry.name: String
    get() = when (this) {
        is NPProtocolEntry.Single -> protocol.name
        is NPProtocolEntry.Composite -> composite.name
        is NPProtocolEntry.Limits -> limits.name
    }

val NPProtocolEntry.isReadOnly: Boolean
    get() = when (this) {
        is NPProtocolEntry.Single -> protocol.isReadOnly
        is NPProtocolEntry.Composite -> composite.isReadOnly
        is NPProtocolEntry.Limits -> false
    }

val NPProtocolEntry.isComposite: Boolean get() = this is NPProtocolEntry.Composite

val NPProtocolDefinition.requiredModalityTypes: Set<NPModalityType>
    get() = modalities.filter { it.enabled }.map { it.modalityType }.toSet()

fun NPProtocolDefinition.requires1064SmartModule(): Boolean =
    modalities.filter { it.enabled }.any {
        val p = it.params
        p is NPModalityParams.PbmTranscranial && p.params.wavelength.requiresSmartModule
    }

// MARK: - Availability ------------------------------------------------------------------

enum class DeviceTier { NONE, T1, T2 }

/** Port of iOS ProtocolAvailability, incl. the ISC-90 `.eegConsentRequired` case. */
sealed class ProtocolAvailability {
    object Available : ProtocolAvailability()
    data class Unavailable(val missingModalities: List<NPModalityType>) : ProtocolAvailability()
    object EegConsentRequired : ProtocolAvailability()

    val isAvailable: Boolean get() = this is Available
}

// MARK: - NPProtocolLibrary -------------------------------------------------------------

class NPProtocolLibrary(
    private val kv: KeyValueStore,
) {
    companion object {
        const val USER_PROTOCOLS_KEY = "np.user-protocols.npps"
        const val BIPA_ACCEPTED_KEY = "np.onboarding.bipa-accepted"
    }

    /** BIPA biometric consent (ISC-90). Seeded from the shared onboarding key. */
    var eegConsentGranted: Boolean = kv.getBoolean(BIPA_ACCEPTED_KEY)
        private set

    var connectedDeviceTier: DeviceTier = DeviceTier.NONE
        private set
    var has1064SmartModules: Boolean = false
        private set
    var availableModalities: Set<NPModalityType> = NP_T1_MODALITY_SET
        private set

    val bundledProtocols: List<NPProtocolEntry> = loadBundledProtocols()

    private val _userProtocols: MutableList<NPProtocolEntry> = mutableListOf()
    val userProtocols: List<NPProtocolEntry> get() = _userProtocols

    val allProtocols: List<NPProtocolEntry> get() = bundledProtocols + _userProtocols

    private val validationCache: MutableMap<UUID, NPValidationResult> = mutableMapOf()

    init {
        loadFromStore()
    }

    // MARK: Availability

    fun availability(entry: NPProtocolEntry): ProtocolAvailability {
        // BIPA gate (ISC-90) runs before hardware/tier — parity with iOS.
        if (!eegConsentGranted && ProtocolConsentGate.isEEGDependent(entry, ::findSingle)) {
            return ProtocolAvailability.EegConsentRequired
        }
        if (connectedDeviceTier == DeviceTier.NONE) {
            return ProtocolAvailability.Unavailable(emptyList())
        }
        return when (entry) {
            is NPProtocolEntry.Composite -> {
                val missing = linkedSetOf<NPModalityType>()
                for (layer in entry.composite.layers) {
                    val def = findSingle(layer.protocolName) ?: continue
                    val sub = availability(NPProtocolEntry.Single(def))
                    if (sub is ProtocolAvailability.Unavailable) missing.addAll(sub.missingModalities)
                }
                if (missing.isEmpty()) ProtocolAvailability.Available
                else ProtocolAvailability.Unavailable(missing.toList())
            }
            is NPProtocolEntry.Single -> {
                val missing = mutableListOf<NPModalityType>()
                val def = entry.protocol
                if (def.requires1064SmartModule() && !has1064SmartModules) {
                    missing.add(NPModalityType.PBM_TRANSCRANIAL)
                }
                for (modality in def.requiredModalityTypes) {
                    if (!isModalityAvailable(modality) && !missing.contains(modality)) missing.add(modality)
                }
                if (missing.isEmpty()) ProtocolAvailability.Available
                else ProtocolAvailability.Unavailable(missing)
            }
            is NPProtocolEntry.Limits -> ProtocolAvailability.Available
        }
    }

    private fun isModalityAvailable(modality: NPModalityType): Boolean = when (modality.tier) {
        NPModalityTier.T1 -> connectedDeviceTier == DeviceTier.T1 || connectedDeviceTier == DeviceTier.T2
        NPModalityTier.T2 -> connectedDeviceTier == DeviceTier.T2
        NPModalityTier.ACCESSORY -> true
    }

    fun updateAvailability(tier: DeviceTier, smartModules: Boolean) {
        connectedDeviceTier = tier
        has1064SmartModules = smartModules
        availableModalities = when (tier) {
            DeviceTier.NONE -> emptySet()
            DeviceTier.T1 -> NP_T1_MODALITY_SET
            DeviceTier.T2 -> NP_T2_MODALITY_SET
        }
    }

    /** Push the BIPA consent state from onboarding (ISC-90), parity with iOS updateEEGConsent. */
    fun updateEEGConsent(granted: Boolean) {
        eegConsentGranted = granted
    }

    // MARK: Validation integration

    fun validationResult(
        entry: NPProtocolEntry,
        limits: NPLimitsSet = NPLimitsSet(name = "unlimited"),
    ): NPValidationResult {
        validationCache[entry.id]?.let { return it }
        val result = NPProtocolValidator(limits).validate(entry, ::findSingle)
        validationCache[entry.id] = result
        return result
    }

    fun issueCount(entry: NPProtocolEntry): Pair<Int, Int> {
        val result = validationCache[entry.id] ?: return 0 to 0
        return result.errors.size to result.warnings.size
    }

    fun revalidateAll(limits: NPLimitsSet = NPLimitsSet(name = "unlimited")) {
        validationCache.clear()
        val validator = NPProtocolValidator(limits)
        for (entry in allProtocols) validationCache[entry.id] = validator.validate(entry, ::findSingle)
    }

    // MARK: Read-only guards

    fun canEdit(entry: NPProtocolEntry): Boolean = !entry.isReadOnly
    fun canDelete(entry: NPProtocolEntry): Boolean = !entry.isReadOnly

    // MARK: Mutations

    fun save(entry: NPProtocolEntry) {
        if (bundledProtocols.any { it.id == entry.id }) return // never overwrite a bundled entry
        val idx = _userProtocols.indexOfFirst { it.id == entry.id }
        if (idx >= 0) _userProtocols[idx] = entry else _userProtocols.add(entry)
        saveToStore()
    }

    fun delete(id: UUID) {
        if (bundledProtocols.any { it.id == id }) return
        _userProtocols.removeAll { it.id == id }
        saveToStore()
    }

    // MARK: Script export / import

    fun exportScript(entry: NPProtocolEntry): String = NPPSSerializer().serialize(entry)

    fun importScript(text: String): NPProtocolEntry {
        val entries = NPPSParser(NPPSLexer(text).tokenize()).parse()
        return entries.firstOrNull() ?: throw NPPSError("No protocol found in script", 0)
    }

    // MARK: Lookup

    private fun findSingle(name: String): NPProtocolDefinition? =
        allProtocols.firstNotNullOfOrNull { e ->
            (e as? NPProtocolEntry.Single)?.protocol?.takeIf { it.name == name }
        }

    // MARK: Bundled loading

    private fun loadBundledProtocols(): List<NPProtocolEntry> =
        NPBundledProtocols.allContents.flatMap { content ->
            runCatching { NPPSParser(NPPSLexer(content).tokenize()).parse() }.getOrDefault(emptyList())
        }

    // MARK: Persistence (NPPS text via KeyValueStore)

    private fun loadFromStore() {
        val text = kv.getString(USER_PROTOCOLS_KEY) ?: return
        if (text.isBlank()) return
        val parsed = runCatching { NPPSParser(NPPSLexer(text).tokenize()).parse() }.getOrDefault(emptyList())
        _userProtocols.clear()
        _userProtocols.addAll(parsed)
    }

    private fun saveToStore() {
        val serializer = NPPSSerializer()
        val text = _userProtocols.joinToString("\n\n") { serializer.serialize(it) }
        kv.putString(USER_PROTOCOLS_KEY, text)
    }
}
